### PR TITLE
READY message

### PR DESCRIPTION
**Pull request by jerch**
As described [here](https://github.com/xtermjs/xterm-addon-attach/pull/9#issuecomment-492263090) there is a small chance that an early sent message from the server is lost due to the client not being ready.

This PR introduces an optional flag `sendReady` for the constructor to instruct the addon to send a message `'#READY#'` when it is done with the initialization. Once the server sees this message it is save to go on with normal messaging.

For remote websocket transport on a newly created socket the flag should always be set. (The flag defaults to `false` due to additional preparations needed on server side).

### CODE REVIEW

File: typings/attach.d.ts [hunk 1]
```diff
@@ -16,15 +16,23 @@ export interface IAttachOptions {
...
-  inputUtf8?: boolean
+  inputUtf8?: boolean;
+
+  /**
+   * Whether to send a '#READY#' message upon activation.
+   * If set the server shall not send data before it has seen this message.
+   * Always set this to avoid losing the first messages.
+   * Since this options needs additional preparations on server side it defaults to `false`.
+   */
+  sendReady?: boolean;
```
Comment (Tyriar) [inline typings/attach.d.ts:35]: Is this common practice when dealing with websockets or something? I would have thought this would be handled by the connection stuff, any links that explain the problem in more detail?

Comment (jerch) [reply typings/attach.d.ts:35]: Well the problem is on our side or how the attach addon sets up the terminal connection  with a websocket object:
- user creates websocket: connection gets established and upgraded in parallel (in browser C++ realms)
- attach addon ctor is called with that instance some times later: by that time the actual state may be still `CONNECTING` or already `OPEN` (or `CLOSING/CLOSED` upon errors)
- if the websocket can be established very fast, the demo server starts sending chunks right away, but the attach addon might not yet have registered the terminal callbacks - lost messages

I see this behavior in the puppeteer tests with this code:
```TS
const ws = new WebSocket('ws://localhost:${port}');
const attachAddon = new window.AttachAddon(ws);  // to late to see all messages
window.term.loadAddon(attachAddon);  // to late
```

The problem is the time span from `new WebSocket` to `loadAddon` with a server not waiting for the latter action, thus sending data right away while no client listener is attached yet. With a READY message a server can explicitly wait for that. Its like an additional ACK telling the server "yes I am ready to consume pty chunks".

I was not able to find reliable information about this particular behavior. The best I found was [this](https://stackoverflow.com/a/34900879), but the marked answer is kinda wrong since it does not reflect the behavior I see with the puppeteer tests (it shows the behavior of answer 2). Also the parts he is referring to are not in the active [spec](https://html.spec.whatwg.org/multipage/web-sockets.html#handler-websocket-onmessage) anymore.

Edit:
Reading answer 2 with its comments more closely it actually might be a result of `addEventListener` vs. `onmessage`. Have to test this.

Edit2:
Ok I know whats going on. The separate `evaluate` calls here:
```TS
await page.evaluate(`const ws = new WebSocket('ws://localhost:${port}')`);
await page.evaluate(`const attachAddon = new window.AttachAddon(ws, {sendReady: true})`);
await page.evaluate(`window.term.loadAddon(attachAddon)`);
```

are inserted between event loop calls, thus translation more like this:

```TS
setTimeout(function(){this.ws = new WebSocket('ws://localhost:${port}')}, 0);
setTimeout(function(){this.attachAddon = new window.AttachAddon(ws, {sendReady: true})}, 0);
setTimeout(function(){window.term.loadAddon(attachAddon)}, 0);
```

**TL;DR**
Its a puppeteer `evaluate` thing translating into calls with event loop invocations in between. Early websocket messages dont drop unless the event loop gets called.

**Comment by jerch** [comment]
@Tyriar Still have to do final tests, but I think we dont need this as its a problem on our side (how we do the `page.evaluate` calls).

In general websocket user should be aware of the fact, that the WebSocket ctor and the event handler attachment should happen without event loop invocation in between to avoid losing early messages:
```JS
const ws = new Websocket(...);
... // setting up other stuff here is fine
... // as long as the event loop does not run
ws.onmessage = data => {
  // no drops here
}

setTimeout(() => {
  ws.onmessage = data => {
    // early messages scheduled on the event loop
    // before this onmessage binding are lost
  }
}, 0);
```

Since thats a common pattern in JS I dont think we need this READY message safety (up to ppl to do this in more complicated websocket setups reusing resources etc). Our puppeteer tests can easily be fixed by joining those lines into one `evaluate` call.

**Comment by jerch** [comment]
@Tyriar Its safe to assume to catch all messages if no event loop was invoked in between.

Tested with this rather long taking synchronous code (17s on my machine):
```TS
      const ws = new WebSocket('ws://localhost:${port}');
      const attachAddon = new AttachAddon(ws, {inputUtf8: true});
      function fibonacci(num) {
        if (num <= 1) return 1;
        return fibonacci(num - 1) + fibonacci(num - 2);
      }
      console.log(fibonacci(45));
      term.loadAddon(attachAddon);
```
No message drops, also all messages are correctly delivered as typed array.

Closing the PR due to not being an issue with the right test setup.
